src/components/KeyboardCompatibleView/KeyboardCompatibleView.js

```diff
@@ -20,63 +25,96 @@ import { KeyboardContext } from '../../context';
-    }).start();
-  }, [
-    channelHeight,
-    heightAnim,
-    keyboardDismissAnimationDuration,
-    keyboardOpenAnimationDuration,
-  ]);
-
-  const dismissKeyboard = useCallback(() => {
-    Keyboard.dismiss();
-
-    return new Promise((resolve) => {
-      if (!isKeyboardOpen) {
-        // If channel height is already at full length, then don't do anything.
-        resolve();
+    const onKeyboardChange = (e) => {
+      if (!enabled) {
+        return;
+      }
+
+      if (e === null) {
+        setChannelHeight(initialHeight.current);
+        isKeyboardOpen.current = false;
+        return;
+      }
+
+      const { endCoordinates } = e;
+      let isOpening = true;
+
+      if (Platform.OS === 'ios') {
+        const { startCoordinates } = e;
+        isOpening = endCoordinates.screenY < startCoordinates.screenY;
+      }
+
+      const keyboardY = endCoordinates.screenY;
+      if (rootChannelView) {
+        if (e.duration && e.easing) {
+          LayoutAnimation.configureNext({
+            // We have to pass the duration equal to minimal accepted duration defined here: RCTLayoutAnimation.m
+            duration: e.duration > 10 ? e.duration : 10,
+            update: {
+              duration: e.duration > 10 ? e.duration : 10,
+              type: LayoutAnimation.Types[e.easing] || 'keyboard',
+            },
+          });
+        }
+        rootChannelView.current.measureInWindow((x, y) => {
+          if (Platform.OS === 'android') {
+            setChannelHeight(
+              isOpening
+                ? keyboardY - y - StatusBar.currentHeight
+                : initialHeight.current,
+            );
+          } else {
+            setChannelHeight(isOpening ? keyboardY - y : initialHeight.current);
+          }
+          isKeyboardOpen.current = !!isOpening;
+        });
+      }
+    };
+
+    let subscriptions = [];
+    if (appState === 'active' && enabled) {
+      if (Platform.OS === 'ios') {
+        subscriptions = [
+          Keyboard.addListener('keyboardWillChangeFrame', onKeyboardChange),
+        ];
       } else {
-        // Bring the channel height to its full length state.
-        Animated.timing(heightAnim, {
-          duration: keyboardDismissAnimationDuration,
-          toValue: initialHeight,
-          useNativeDriver: false,
-        }).start(resolve);
+        subscriptions = [
+          Keyboard.addListener('keyboardDidHide', onKeyboardChange),
+          Keyboard.addListener('keyboardDidShow', onKeyboardChange),
+        ];
+      }
+    }
+    return () => {
+      subscriptions.forEach((subscription) => {
+        subscription.remove();
+      });
+    };
+  }, [enabled, appState]);
+
+  const dismissKeyboard = useCallback((callback) => {
+    if (!isKeyboardOpen.current) {
+      callback && callback();
+    } else {
+      if (callback) {
+        const subscribe = Keyboard.addListener('keyboardDidHide', () => {
+          callback();
+          subscribe.remove();
```

Comment:
so the subscribe const being set as this listener is being used within the callback of the same listener for the cleanup? does that work? seems a bit circular